src/server/command/worldedit/pattern.zig

```diff
@@ -0,0 +1,58 @@
+const std = @import("std");
+
+const main = @import("root");
+const AliasTable = main.utils.AliasTable;
+const Block = main.blocks.Block;
+const ListUnmanaged = main.ListUnmanaged;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+blocks: AliasTable(Entry),
+
+const Entry = struct {
+	block: Block,
+	chance: f32,
+};
+
+pub fn initFromString(allocator: NeverFailingAllocator, source: []const u8) !@This() {
+	var specifiers = std.mem.splitScalar(u8, source, ',');
+	var totalWeight: usize = 0;
+
+	var weightedEntries: ListUnmanaged(struct {block: Block, weight: usize}) = .{};
+	defer weightedEntries.deinit(main.stackAllocator);
+
+	while(specifiers.next()) |specifier| {
+		var iterator = std.mem.splitScalar(u8, specifier, '%');
+
+		const first = iterator.next();
+		const second = iterator.next();
+
+		if(iterator.peek() != null) {
+			return error.PatternSyntaxError;
+		}
+		if(second) |blockId| {
+			if(first) |weight| {
+				const block = main.blocks.parseBlock(blockId);
+				const blockWeight = try std.fmt.parseInt(usize, weight, 10);
```

Comment:
Shouldn't we allow for floats? I think specifying 0.1% is a valid use-case.

Reply:
It can be done by setting weight of other blocks to bigger value. 

`0.1%cubyz:stone,1%cubyz:oak_planks` is equivalent to `1%cubyz:stone,10%cubyz:oak_planks` since chances are normalized.

Reply:
Yeah, I know, but it could be rather tedious if you already set all the other values.
Also this interaction may not be as obvious to the average player.